codegen-renderers/src/main/kotlin/io/vrap/codegen/languages/typescript/joi/JoiValidatorModuleRenderer.kt

```diff
@@ -0,0 +1,185 @@
+package io.vrap.codegen.languages.typescript.joi
+
+import com.google.inject.Inject
+import io.vrap.codegen.languages.extensions.hasSubtypes
+import io.vrap.codegen.languages.java.extensions.EObjectTypeExtensions
+import io.vrap.codegen.languages.java.extensions.simpleName
+import io.vrap.rmf.codegen.io.TemplateFile
+import io.vrap.rmf.codegen.rendring.FileProducer
+import io.vrap.rmf.codegen.rendring.utils.keepIndentation
+import io.vrap.rmf.codegen.types.*
+import io.vrap.rmf.raml.model.types.*
+
+class JoiValidatorModuleRenderer @Inject constructor(override val vrapTypeProvider: VrapTypeProvider) : JoiObjectTypeExtensions, EObjectTypeExtensions, FileProducer {
+
+    @Inject
+    lateinit var allAnyTypes: MutableList<AnyType>
+
+
+    override fun produceFiles(): List<TemplateFile> {
+        return allAnyTypes.groupBy { s(it) }
+            .map { entry: Map.Entry<String, List<AnyType>> -> buildModule("${entry.key}-types", entry.value) }
+            .toList()
+    }
+
+    private fun s(it: AnyType): String {
+        val t = it.toVrapType()
+        return when (t) {
+            is VrapObjectType -> t.`package`
+            is VrapEnumType -> t.`package`
+            else -> ""
+        }
+    }
+
+
+    fun buildModule(moduleName: String, types: List<AnyType>): TemplateFile {
+
+        val content = """
+           |/* tslint:disable */
+           |//Generated file, please do not change
+           |
+           |import * as Joi from 'joi'
+           |${getImportsForJoiModule(moduleName, types)}
+           |
+           |${types.map { it.renderAnyType() }.joinToString(separator = "\n\n")}
+       """.trimMargin().keepIndentation()
+
+        return TemplateFile(content, moduleName.replace(".", "/") + ".ts")
+
+    }
+
+    fun AnyType.renderAnyType(): String {
+        return when (this) {
+            is ObjectType -> this.renderObjectType()
+            is StringType -> this.renderStringType()
+            else -> throw IllegalArgumentException("unhandled case ${this.javaClass}")
+
+        }
+    }
+
+    fun ObjectType.renderObjectType(): String {
+        val vrapType = this.toVrapType() as VrapObjectType
+        if (this.hasSubtypes()) {
+
+            val allSubsCases = this.subTypes
+                .filterIsInstance(ObjectType::class.java)
+                .filter { !it.discriminatorValue.isNullOrEmpty() }
+                .map {
+                    """
+                    |.when('${this.discriminator}',{
+                    |    is: '${it.discriminatorValue}',
+                    |    then: ${it.toVrapType().simpleJoiName()}()
+                    |})
+                    """.trimMargin()
+                }.joinToString(separator = "\n")
+
+            return """
+            |export function ${vrapType.simpleJoiName()}(){
+            |   return Joi.alternatives()
```

Comment:
Very nice 🥇 This makes the generated code more powerful than our current manually created code 😄 😎